panic: Calling IP() on a DomainAddress.

<!--
除非特殊情况，请完整填写所有问题。不按模板发的 issue 将直接被关闭。
如果你遇到的问题不是 V2Ray 的 bug，比如你不清楚如何配置，请在 https://github.com/v2fly/v2ray-core/discussions 进行讨论。
-->

## 你正在使用哪个版本的 V2Ray？

<!-- 如果服务端和客户端使用了不同版本，请注明 -->
服务端：4.38.3
客户端：4.32.1


## 你的使用场景是什么？

<!-- 比如使用 Chrome 通过 Socks/VMess 代理观看 YouTube 视频 -->

使用手机端v2rayNG绕过大陆及局域网地址代理


## 你看到的异常现象是什么？

<!-- 请描述具体现象，比如访问超时、TLS 证书错误等 -->

v2ray -config /usr/local/etc/v2ray/config.json
V2Ray 4.38.3 (V2Fly, a community-driven edition of V2Ray.) Custom (go1.16.3 linux/amd64)
A unified platform for anti-censorship.
2021/05/01 03:42:20 [Info] main/jsonem: Reading config: /usr/local/etc/v2ray/config.json
panic: Calling IP() on a DomainAddress.

goroutine 1 [running]:
github.com/v2fly/v2ray-core/v4/common/net.domainAddress.IP(...)
        github.com/v2fly/v2ray-core/v4/common/net/address.go:164
github.com/v2fly/v2ray-core/v4/app/browserforwarder.(*Forwarder).Start(0xc0006886c0, 0x0, 0x0)
        github.com/v2fly/v2ray-core/v4/app/browserforwarder/forwarder.go:57 +0x127
github.com/v2fly/v2ray-core/v4.(*Instance).Start(0xc0005cdc70, 0x0, 0x0)
        github.com/v2fly/v2ray-core/v4/v2ray.go:337 +0xac
main.main()
        github.com/v2fly/v2ray-core/v4/main/main.go:151 +0x15a


## 你期待看到的正常表现是怎样的？

正常启动

## 请附上你的配置

<!-- 提交 issue 前，请隐去服务器域名或 IP 地址 -->

**服务端配置：**

```javascript
// 在这里附上服务器端配置文件
{
	"log": {
		"access": "/var/log/v2ray/access.log",
		"error": "/var/log/v2ray/error.log",
		"loglevel": "debug"
	},
	"dns": {
		"servers": [
			"localhost"
		],
		"clientIp": "1.2.3.4"
	},
	"routing": {
		"domainStrategy": "IPIfNonMatch",
		"domainMatcher": "mph",
		"rules": [{
				"type": "field",
				"ip": [
					"0.0.0.0/8",
					"10.0.0.0/8",
					"100.64.0.0/10",
					"127.0.0.0/8",
					"169.254.0.0/16",
					"172.16.0.0/12",
					"192.0.0.0/24",
					"192.0.2.0/24",
					"192.168.0.0/16",
					"198.18.0.0/15",
					"198.51.100.0/24",
					"203.0.113.0/24",
					"::1/128",
					"fc00::/7",
					"fe80::/10"
				],
				"outboundTag": "blocked"
			},
			{
				"type": "field",
				"outboundTag": "blocked",
				"protocol": [
					"bittorrent"
				]
			},
			{
				"type": "field",
				"domain": ["geosite:category-ads"],
				"outboundTag": "blocked"
			}
		],
		"balancers": []
	},
	"policy": {},
	"inbounds": [{
			"listen": "/var/tmp/v2ray/vmess.sock",
			"protocol": "vmess",
			"settings": {
				"clients": [{
						"id": "（id1已隐去）",
						"alterId": 64
					},
					{
						"id": "（id2已隐去）",
						"alterId": 16
					}
				]
			},
			"streamSettings": {
				"network": "ws",
				"wsSettings": {
					"acceptProxyProtocol": true,
					"path": "/path_1"
				}
			}
		},
		{
			"listen": "/var/tmp/v2ray/vless.sock",
			"protocol": "vless",
			"settings": {
				"clients": [{
						"id": "（id1已隐去）"
					},
					{
						"id": "（id2已隐去）"
					}
				],
				"decryption": "none"
			},
			"streamSettings": {
				"network": "ws",
				"wsSettings": {
					"acceptProxyProtocol": true,
					"path": "/path_2"
				}
			}
		}
	],
	"outbounds": [{
			"protocol": "freedom",
			"settings": {}
		},
		{
			"protocol": "blackhole",
			"settings": {},
			"tag": "blocked"
		}
	],
	"transport": {},
	"stats": {},
	"reverse": {},
	"fakedns": [],
	"browserForwarder": {},
	"observatory": {}
}

```

**客户端配置：**

```javascript
// 在这里附上客户端配置
{
	"inbounds": [{
			"port": 10809,
			"listen": "0.0.0.0",
			"protocol": "http",
			"settings": {
				"timeout": 0
			},
			"tag": "proxy-inbound-http"
		},
		{
			"port": 10808,
			"listen": "0.0.0.0",
			"protocol": "socks",
			"settings": {
				"ip": "127.0.0.1",
				"auth": "noauth",
				"udp": true
			},
			"tag": "proxy-inbound-socks"
		}
	],
	"outbounds": [{
			"protocol": "vmess",
			"settings": {
				"vnext": [{
					"address": "地址已隐去",
					"port": 443,
					"users": [{
						"id": "id已隐去",
						"alterId": 16
					}]
				}]
			},
			"streamSettings": {
				"network": "ws",
				"security": "tls",
				"wsSettings": {
					"path": "/path_1",
					"headers": {
						"Host": "example.com"
					}
				}
			},
			"mux": {
				"enabled": true
			},
			"tag": "proxy"
		},
		{
			"protocol": "freedom",
			"settings": {

			},
			"tag": "direct"
		},
		{
			"protocol": "blackhole",
			"settings": {

			},
			"tag": "blocked"
		}
	],
	"routing": {
		"domainStrategy": "IPOnDemand",
		"rules": [{
				"type": "field",
				"outboundTag": "proxy",
				"domain": [
					"services.google.cn"
				]
			},
			{
				"type": "field",
				"ip": [
					"geoip:private"
				],
				"outboundTag": "direct"
			},
			{
				"type": "field",
				"ip": [
					"geoip:cn"
				],
				"outboundTag": "direct"
			},
			{
				"type": "field",
				"domain": [
					"geosite:cn"
				],
				"outboundTag": "direct"
			},
			{
				"type": "field",
				"outboundTag": "direct",
				"protocol": [
					"bittorrent"
				]
			},
			{
				"type": "field",
				"domain": [
					"geosite:category-ads"
				],
				"outboundTag": "blocked"
			}
		]
	}
}
```

## 请附上出错时软件输出的错误日志

<!-- 在 Linux 中，日志通常在 `/var/log/v2ray/error.log` 文件中 -->

**服务器端错误日志：**

```javascript
// 在这里附上服务器端日志
2021/05/01 03:57:55 [Debug] app/log: Logger started
2021/05/01 03:57:55 [Info] app/dns: DNS: created localhost client
2021/05/01 03:57:55 [Info] app/dns: DNS: client localhost uses clientIP 1.2.3.4
2021/05/01 03:57:55 [Debug] app/router: MphDomainMatcher is enabled for 616 domain rule(s)
2021/05/01 03:57:55 [Debug] app/proxyman/inbound: creating unix domain socket worker on /var/tmp/v2ray/vmess.sock
2021/05/01 03:57:55 [Debug] app/proxyman/inbound: creating unix domain socket worker on /var/tmp/v2ray/vless.sock
2021/05/01 03:57:55 [Info] transport/internet/websocket: listening unix domain socket(for WS) on /var/tmp/v2ray/vmess.sock
2021/05/01 03:57:55 [Warning] transport/internet/websocket: accepting PROXY protocol
2021/05/01 03:57:55 [Info] transport/internet/websocket: listening unix domain socket(for WS) on /var/tmp/v2ray/vless.sock
```

**客户端错误日志：**

```javascript
// 在这里附上客户端日志
（服务端无法启动）
```

## 请附上访问日志

<!-- 在 Linux 中，访问日志通常在 `/var/log/v2ray/access.log` 文件中 -->

```javascript
// 在这里附上服务器端日志

```

## 其它相关的配置文件（如 Nginx）和相关日志
Nginx - /etc/nginx/nginx.conf
（仅摘录proxy部分）

        location /path_1 {
        	proxy_redirect off;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
            proxy_http_version 1.1;
            proxy_read_timeout 300s;
            proxy_send_timeout 300s;
            proxy_pass http://unix:/var/tmp/v2ray/vmess.sock:;
        }
        
        location /path_2 {
        proxy_redirect off;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
            proxy_http_version 1.1;
            proxy_read_timeout 300s;
            proxy_send_timeout 300s;
            proxy_pass http://unix:/var/tmp/v2ray/vless.sock:;
        }

## 如果 V2Ray 无法启动，请附上 `--test` 命令的输出

V2Ray 4.38.3 (V2Fly, a community-driven edition of V2Ray.) Custom (go1.16.3 linux/amd64)
A unified platform for anti-censorship.
2021/05/01 04:02:43 [Info] main/jsonem: Reading config: /usr/local/etc/v2ray/config.json
Configuration OK.

## 如果 V2Ray 服务运行异常，请附上 journal 日志

<!-- 通常的命令为 `journalctl -u v2ray` -->

<!-- 请预览一下你填写的内容并整理好格式后，再提交 -->

--- a/app/browserforwarder/forwarder.go
+++ b/app/browserforwarder/forwarder.go
@@ -64,7 +64,7 @@ func (f *Forwarder) Start() error {
 		} else if address.Domain() == "localhost" {
 			listener, err = internet.ListenSystem(f.ctx, &net.TCPAddr{IP: net.IP{127, 0, 0, 1}, Port: int(f.config.ListenPort)}, nil)
 		} else {
-			return newError("cannot listen on the address")
+			return newError("forwarder cannot listen on the address")
 		}
 
 		if err != nil {

--- a/app/browserforwarder/forwarder.go
+++ b/app/browserforwarder/forwarder.go
@@ -54,8 +54,19 @@ func (f *Forwarder) Start() error {
 	if f.config.ListenAddr != "" {
 		f.forwarder = handler.NewHttpHandle()
 		f.httpserver = &http.Server{Handler: f}
+
+		var listener net.Listener
+		var err error
 		address := net.ParseAddress(f.config.ListenAddr)
-		listener, err := internet.ListenSystem(f.ctx, &net.TCPAddr{IP: address.IP(), Port: int(f.config.ListenPort)}, nil)
+
+		if address.Family().IsIP() {
+			listener, err = internet.ListenSystem(f.ctx, &net.TCPAddr{IP: address.IP(), Port: int(f.config.ListenPort)}, nil)
+		} else if address.Domain() == "localhost" {
+			listener, err = internet.ListenSystem(f.ctx, &net.TCPAddr{IP: net.IP{127, 0, 0, 1}, Port: int(f.config.ListenPort)}, nil)
+		} else {
+			return newError("cannot listen on the address")
+		}
+
 		if err != nil {
 			return newError("forwarder cannot listen on the port").Base(err)
 		}

--- a/app/browserforwarder/forwarder.go
+++ b/app/browserforwarder/forwarder.go
@@ -59,11 +59,12 @@ func (f *Forwarder) Start() error {
 		var err error
 		address := net.ParseAddress(f.config.ListenAddr)
 
-		if address.Family().IsIP() {
+		switch {
+		case address.Family().IsIP():
 			listener, err = internet.ListenSystem(f.ctx, &net.TCPAddr{IP: address.IP(), Port: int(f.config.ListenPort)}, nil)
-		} else if address.Domain() == "localhost" {
+		case address.Domain() == "localhost":
 			listener, err = internet.ListenSystem(f.ctx, &net.TCPAddr{IP: net.IP{127, 0, 0, 1}, Port: int(f.config.ListenPort)}, nil)
-		} else {
+		default:
 			return newError("forwarder cannot listen on the address")
 		}
 

--- a/app/browserforwarder/forwarder.go
+++ b/app/browserforwarder/forwarder.go
@@ -51,19 +51,21 @@ func (f *Forwarder) Type() interface{} {
 }
 
 func (f *Forwarder) Start() error {
-	f.forwarder = handler.NewHttpHandle()
-	f.httpserver = &http.Server{Handler: f}
-	address := net.ParseAddress(f.config.ListenAddr)
-	listener, err := internet.ListenSystem(f.ctx, &net.TCPAddr{IP: address.IP(), Port: int(f.config.ListenPort)}, nil)
-	if err != nil {
-		return newError("forwarder cannot listen on the port").Base(err)
-	}
-	go func() {
-		err = f.httpserver.Serve(listener)
+	if f.config.ListenAddr != "" {
+		f.forwarder = handler.NewHttpHandle()
+		f.httpserver = &http.Server{Handler: f}
+		address := net.ParseAddress(f.config.ListenAddr)
+		listener, err := internet.ListenSystem(f.ctx, &net.TCPAddr{IP: address.IP(), Port: int(f.config.ListenPort)}, nil)
 		if err != nil {
-			newError("cannot serve http forward server").Base(err).WriteToLog()
+			return newError("forwarder cannot listen on the port").Base(err)
 		}
-	}()
+		go func() {
+			err = f.httpserver.Serve(listener)
+			if err != nil {
+				newError("cannot serve http forward server").Base(err).WriteToLog()
+			}
+		}()
+	}
 	return nil
 }
 

--- a/app/browserforwarder/forwarder.go
+++ b/app/browserforwarder/forwarder.go
@@ -7,6 +7,7 @@ import (
 	"context"
 	"io"
 	"net/http"
+	"strings"
 	"time"
 
 	"github.com/v2fly/BrowserBridge/handler"
@@ -62,18 +63,17 @@ func (f *Forwarder) Start() error {
 		switch {
 		case address.Family().IsIP():
 			listener, err = internet.ListenSystem(f.ctx, &net.TCPAddr{IP: address.IP(), Port: int(f.config.ListenPort)}, nil)
-		case address.Domain() == "localhost":
+		case strings.EqualFold(address.Domain(), "localhost"):
 			listener, err = internet.ListenSystem(f.ctx, &net.TCPAddr{IP: net.IP{127, 0, 0, 1}, Port: int(f.config.ListenPort)}, nil)
 		default:
-			return newError("forwarder cannot listen on the address")
+			return newError("forwarder cannot listen on the address: ", address)
+		}
+		if err != nil {
+			return newError("forwarder cannot listen on the port ", f.config.ListenPort).Base(err)
 		}
 
-		if err != nil {
-			return newError("forwarder cannot listen on the port").Base(err)
-		}
 		go func() {
-			err = f.httpserver.Serve(listener)
-			if err != nil {
+			if err := f.httpserver.Serve(listener); err != nil {
 				newError("cannot serve http forward server").Base(err).WriteToLog()
 			}
 		}()

--- a/infra/conf/browser_forwarder.go
+++ b/infra/conf/browser_forwarder.go
@@ -1,6 +1,8 @@
 package conf
 
 import (
+	"strings"
+
 	"github.com/golang/protobuf/proto"
 
 	"github.com/v2fly/v2ray-core/v4/app/browserforwarder"
@@ -11,7 +13,11 @@ type BrowserForwarderConfig struct {
 	ListenPort int32  `json:"listenPort"`
 }
 
-func (b BrowserForwarderConfig) Build() (proto.Message, error) {
+func (b *BrowserForwarderConfig) Build() (proto.Message, error) {
+	b.ListenAddr = strings.TrimSpace(b.ListenAddr)
+	if b.ListenAddr != "" && b.ListenPort == 0 {
+		b.ListenPort = 54321
+	}
 	return &browserforwarder.Config{
 		ListenAddr: b.ListenAddr,
 		ListenPort: b.ListenPort,

